### PR TITLE
POC httproute priorities

### PR DESCRIPTION
**Pull request by mlavacca**
> NOTE: **This PR is to be intended as a POC and will not be merged.**

> NOTE_2: all the code lines commented in this PR are workarounds to compile the code and execute the new test. They are not to be intended as a proposal for the actual implementation

**What this PR does / why we need it**:

This is a poc for setting priorities to all the HTTPRouteMatches, even when coming from different `HTTPRoutes`.

To do so, the following has been implemented:
1. all the HTTPRoutesMatches have been isolated along with each hostname for each HTTPRoutes and have been like if they were unique `HTTPRoutes`
2. all the `HTTPRoutes` created in the step above are sorted according to the Gateway API specification (pasted below)
3. the HTTPRoutes are inversely iterated and an increasing priority is assigned to them.

A omni-comprehensive test has been added to verify and showcase how the logic works.

### Gateway API specification:

```go
	// In the event that multiple HTTPRoutes specify intersecting hostnames (e.g.
	// overlapping wildcard matching and exact matching hostnames), precedence must
	// be given to rules from the HTTPRoute with the largest number of:
	//
	// * Characters in a matching non-wildcard hostname.
	// * Characters in a matching hostname.
	//
	// If ties exist across multiple Routes, the matching precedence rules for
	// HTTPRouteMatches takes over.
        // ...
	// Proxy or Load Balancer routing configuration generated from HTTPRoutes
	// MUST prioritize matches based on the following criteria, continuing on
	// ties. Across all rules specified on applicable Routes, precedence must be
	// given to the match with the largest number of:
	//
	// * Characters in a matching "Exact" path match
	// * Characters in a matching "Prefix" path match
	// * Header matches.
	// * Query param matches.
	//
	// Note: The precedence of RegularExpression path matches are implementation-specific.
	//
	// If ties still exist across multiple Routes, matching precedence MUST be
	// determined in order of the following criteria, continuing on ties:
	//
	// * The oldest Route based on creation timestamp.
	// * The Route appearing first in alphabetical order by
	//   "{namespace}/{name}".
	//
	// If ties still exist within an HTTPRoute, matching precedence MUST be granted
	// to the FIRST matching rule (in list order) with a match meeting the above
	// criteria.
```

~This implementation takes into account all the above rules but the hostname wildcard one and the latest related to the precedence within `HTTPRoutes`. They are just additional sorting rules that will be taken into account and implemented if the POC is accepted.~

All the specification rules have been implemented.

**Which issue this PR fixes**:

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/translate_httproute.go [hunk 1]
```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
...
+		namespacedNamei := fmt.Sprintf("%s/%s", routes[i].Namespace, routes[i].Name)
+		namespacedNamej := fmt.Sprintf("%s/%s", routes[j].Namespace, routes[j].Name)
+
+		switch {
+		// hostname matching
+		case hasHostnamei && !hasHostnamej:
+			return true
+		case !hasHostnamei && hasHostnamej:
+			return false
+		case hasHostnamei && hasHostnamej && len(routes[i].Spec.Hostnames[0]) != len(routes[j].Spec.Hostnames[0]):
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_httproute.go:657]: The specification did not consider how to compare hostnames when there are multiple hostnames in one `HTTPRoute`?

File: internal/dataplane/parser/translate_httproute.go [hunk 2]
```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
...
+			return len(routes[i].Spec.Hostnames[0]) > len(routes[j].Spec.Hostnames[0])
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_httproute.go:658]: I think we should first compare if one has an exact hostname match but the other has only wildcard hostname match.

File: internal/dataplane/parser/translate_httproute.go [hunk 3]
```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
...
+		// queryparam matching
+		case len(matchi.QueryParams) != len(matchj.QueryParams):
+			return len(matchi.QueryParams) > len(matchj.QueryParams)
+
+		// creation timestamp
+		case routes[i].CreationTimestamp.Compare(routes[j].CreationTimestamp.Time) != 0:
+			return routes[i].CreationTimestamp.After(routes[j].CreationTimestamp.Time)
+
+		// alphabetical order
+		case strings.Compare(namespacedNamei, namespacedNamej) == -1:
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_httproute.go:689]: we can use built in operator `<`  to compare strings here.

File: internal/dataplane/parser/translate_httproute.go [hunk 4]
```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
...
+		// header matching
+		case len(matchi.Headers) != len(matchj.Headers):
+			return len(matchi.Headers) > len(matchj.Headers)
+
+		// queryparam matching
+		case len(matchi.QueryParams) != len(matchj.QueryParams):
+			return len(matchi.QueryParams) > len(matchj.QueryParams)
+
+		// creation timestamp
+		case routes[i].CreationTimestamp.Compare(routes[j].CreationTimestamp.Time) != 0:
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_httproute.go:685]: **Oldest** route should take the highest precedence, so here we should use `time.Before`.

File: internal/dataplane/parser/translate_httproute.go [hunk 5]
```diff
@@ -562,3 +591,133 @@ func httpBackendRefsToBackendRefs(httpBackendRef []gatewayv1beta1.HTTPBackendRef
...
+
+	return newHTTPRoutes
+}
+
+func sortHTTPRoutes(routes []gatewayv1beta1.HTTPRoute) {
+	sort.SliceStable(routes, func(i, j int) bool {
+		hasHostnamei := hasHostnames(routes[i])
+		hasHostnamej := hasHostnames(routes[j])
+
+		matchi := routes[i].Spec.Rules[0].Matches[0]
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_httproute.go:639]: In my comprehension, the sort should run through ALL matches but not only for `HTTPRoutes`. For example, if two `HTTPRoute`s has the following spec:

```
rules:
  - matches:
    - path:  
      value: "/a/b/c"
      type: PathPrefix
    - path:
      value: "/a"
      type: PathPrefix
```
and 
```
rules:
  - matches:
    - path:
      value: "/a/b"
      type: PathPrefix
```
The priority of there matches should be interleaved, `/a/b/c` > `/a/b` >`/a`.

File: internal/dataplane/parser/translate_httproute.go (hunk 4 above)
Comment (mlavacca) [reply internal/dataplane/parser/translate_httproute.go:685]: good catch

File: internal/dataplane/parser/translate_httproute.go (hunk 2 above)
Comment (mlavacca) [reply internal/dataplane/parser/translate_httproute.go:658]: Yes, this check hasn't been implemented here. I wrote it in the PR description. It is straightforward at this point. I'm gonna add it

File: internal/dataplane/parser/translate_httproute.go (hunk 5 above)
Comment (mlavacca) [reply internal/dataplane/parser/translate_httproute.go:639]: Here is your example: https://github.com/Kong/kubernetes-ingress-controller/pull/4020/commits/2bc952a0c76700fc8bd9a6b8e21d3197b4c5f208

File: internal/dataplane/parser/translate_httproute.go (hunk 2 above)
Comment (mlavacca) [reply internal/dataplane/parser/translate_httproute.go:667]: see 67a7cdd1234ff77154622219d97e308d8c976452

**Pull request closed without merging**